// @expo/entity/build/testfixtures/DateIDTestEntity.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DateIDTestEntityPrivacyPolicy = exports.dateIDTestEntityConfiguration = void 0;
const Entity_1 = __importDefault(require("../Entity"));
const EntityConfiguration_1 = __importDefault(require("../EntityConfiguration"));
const EntityFields_1 = require("../EntityFields");
const EntityPrivacyPolicy_1 = __importDefault(require("../EntityPrivacyPolicy"));
const AlwaysAllowPrivacyPolicyRule_1 = __importDefault(require("../rules/AlwaysAllowPrivacyPolicyRule"));
exports.dateIDTestEntityConfiguration = new EntityConfiguration_1.default({
    idField: 'id',
    tableName: 'simple_test_entity_should_not_write_to_db',
    schema: {
        id: new EntityFields_1.DateField({
            columnName: 'custom_id',
        }),
    },
    databaseAdapterFlavor: 'postgres',
    cacheAdapterFlavor: 'redis',
});
class DateIDTestEntityPrivacyPolicy extends EntityPrivacyPolicy_1.default {
    readRules = [
        new AlwaysAllowPrivacyPolicyRule_1.default(),
    ];
    createRules = [
        new AlwaysAllowPrivacyPolicyRule_1.default(),
    ];
    updateRules = [
        new AlwaysAllowPrivacyPolicyRule_1.default(),
    ];
    deleteRules = [
        new AlwaysAllowPrivacyPolicyRule_1.default(),
    ];
}
exports.DateIDTestEntityPrivacyPolicy = DateIDTestEntityPrivacyPolicy;
class DateIDTestEntity extends Entity_1.default {
    static defineCompanionDefinition() {
        return {
            entityClass: DateIDTestEntity,
            entityConfiguration: exports.dateIDTestEntityConfiguration,
            privacyPolicyClass: DateIDTestEntityPrivacyPolicy,
        };
    }
}
exports.default = DateIDTestEntity;
