// @expo/entity/build/testfixtures/TestEntity2.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestEntity2PrivacyPolicy = exports.testEntity2Configuration = void 0;
const Entity_1 = __importDefault(require("../Entity"));
const EntityConfiguration_1 = __importDefault(require("../EntityConfiguration"));
const EntityFields_1 = require("../EntityFields");
const EntityPrivacyPolicy_1 = __importDefault(require("../EntityPrivacyPolicy"));
const AlwaysAllowPrivacyPolicyRule_1 = __importDefault(require("../rules/AlwaysAllowPrivacyPolicyRule"));
exports.testEntity2Configuration = new EntityConfiguration_1.default({
    idField: 'id',
    tableName: 'test_entity_should_not_write_to_db_2',
    schema: {
        id: new EntityFields_1.UUIDField({
            columnName: 'id',
        }),
        foreignKey: new EntityFields_1.UUIDField({
            columnName: 'foreign_key',
        }),
    },
    databaseAdapterFlavor: 'postgres',
    cacheAdapterFlavor: 'redis',
});
class TestEntity2PrivacyPolicy extends EntityPrivacyPolicy_1.default {
    readRules = [
        new AlwaysAllowPrivacyPolicyRule_1.default(),
    ];
    createRules = [
        new AlwaysAllowPrivacyPolicyRule_1.default(),
    ];
    updateRules = [
        new AlwaysAllowPrivacyPolicyRule_1.default(),
    ];
    deleteRules = [
        new AlwaysAllowPrivacyPolicyRule_1.default(),
    ];
}
exports.TestEntity2PrivacyPolicy = TestEntity2PrivacyPolicy;
class TestEntity2 extends Entity_1.default {
    static defineCompanionDefinition() {
        return {
            entityClass: TestEntity2,
            entityConfiguration: exports.testEntity2Configuration,
            privacyPolicyClass: TestEntity2PrivacyPolicy,
        };
    }
}
exports.default = TestEntity2;
