// @expo/entity/build/errors/EntityError.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityErrorCode = exports.EntityErrorState = void 0;
const es6_error_1 = __importDefault(require("es6-error"));
var EntityErrorState;
(function (EntityErrorState) {
    EntityErrorState[EntityErrorState["UNKNOWN"] = 0] = "UNKNOWN";
    EntityErrorState[EntityErrorState["TRANSIENT"] = 1] = "TRANSIENT";
    EntityErrorState[EntityErrorState["PERMANENT"] = 2] = "PERMANENT";
})(EntityErrorState || (exports.EntityErrorState = EntityErrorState = {}));
var EntityErrorCode;
(function (EntityErrorCode) {
    EntityErrorCode["ERR_ENTITY_NOT_AUTHORIZED"] = "ERR_ENTITY_NOT_AUTHORIZED";
    EntityErrorCode["ERR_ENTITY_NOT_FOUND"] = "ERR_ENTITY_NOT_FOUND";
    EntityErrorCode["ERR_ENTITY_INVALID_FIELD_VALUE"] = "ERR_ENTITY_INVALID_FIELD_VALUE";
    EntityErrorCode["ERR_ENTITY_DATABASE_ADAPTER_TRANSIENT"] = "ERR_ENTITY_DATABASE_ADAPTER_TRANSIENT";
    EntityErrorCode["ERR_ENTITY_DATABASE_ADAPTER_UNKNOWN"] = "ERR_ENTITY_DATABASE_ADAPTER_UNKNOWN";
    EntityErrorCode["ERR_ENTITY_DATABASE_ADAPTER_CHECK_CONSTRAINT"] = "ERR_ENTITY_DATABASE_ADAPTER_CHECK_CONSTRAINT";
    EntityErrorCode["ERR_ENTITY_DATABASE_ADAPTER_EXCLUSION_CONSTRAINT"] = "ERR_ENTITY_DATABASE_ADAPTER_EXCLUSION_CONSTRAINT";
    EntityErrorCode["ERR_ENTITY_DATABASE_ADAPTER_FOREIGN_KEY_CONSTRAINT"] = "ERR_ENTITY_DATABASE_ADAPTER_FOREIGN_KEY_CONSTRAINT";
    EntityErrorCode["ERR_ENTITY_DATABASE_ADAPTER_NOT_NULL_CONSTRAINT"] = "ERR_ENTITY_DATABASE_ADAPTER_NOT_NULL_CONSTRAINT";
    EntityErrorCode["ERR_ENTITY_DATABASE_ADAPTER_UNIQUE_CONSTRAINT"] = "ERR_ENTITY_DATABASE_ADAPTER_UNIQUE_CONSTRAINT";
    EntityErrorCode["ERR_ENTITY_CACHE_ADAPTER_TRANSIENT"] = "ERR_ENTITY_CACHE_ADAPTER_TRANSIENT";
})(EntityErrorCode || (exports.EntityErrorCode = EntityErrorCode = {}));
class EntityError extends es6_error_1.default {
    cause;
    constructor(message, cause) {
        super(message);
        this.cause = cause;
    }
}
exports.default = EntityError;
