// @expo/entity/build/errors/EntityDatabaseAdapterError.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityDatabaseAdapterUniqueConstraintError = exports.EntityDatabaseAdapterNotNullConstraintError = exports.EntityDatabaseAdapterForeignKeyConstraintError = exports.EntityDatabaseAdapterExclusionConstraintError = exports.EntityDatabaseAdapterCheckConstraintError = exports.EntityDatabaseAdapterUnknownError = exports.EntityDatabaseAdapterTransientError = void 0;
const EntityError_1 = __importStar(require("./EntityError"));
class EntityDatabaseAdapterError extends EntityError_1.default {
}
exports.default = EntityDatabaseAdapterError;
class EntityDatabaseAdapterTransientError extends EntityDatabaseAdapterError {
    state = EntityError_1.EntityErrorState.TRANSIENT;
    code = EntityError_1.EntityErrorCode.ERR_ENTITY_DATABASE_ADAPTER_TRANSIENT;
}
exports.EntityDatabaseAdapterTransientError = EntityDatabaseAdapterTransientError;
class EntityDatabaseAdapterUnknownError extends EntityDatabaseAdapterError {
    state = EntityError_1.EntityErrorState.UNKNOWN;
    code = EntityError_1.EntityErrorCode.ERR_ENTITY_DATABASE_ADAPTER_UNKNOWN;
}
exports.EntityDatabaseAdapterUnknownError = EntityDatabaseAdapterUnknownError;
class EntityDatabaseAdapterCheckConstraintError extends EntityDatabaseAdapterError {
    state = EntityError_1.EntityErrorState.PERMANENT;
    code = EntityError_1.EntityErrorCode.ERR_ENTITY_DATABASE_ADAPTER_CHECK_CONSTRAINT;
}
exports.EntityDatabaseAdapterCheckConstraintError = EntityDatabaseAdapterCheckConstraintError;
class EntityDatabaseAdapterExclusionConstraintError extends EntityDatabaseAdapterError {
    state = EntityError_1.EntityErrorState.PERMANENT;
    code = EntityError_1.EntityErrorCode.ERR_ENTITY_DATABASE_ADAPTER_EXCLUSION_CONSTRAINT;
}
exports.EntityDatabaseAdapterExclusionConstraintError = EntityDatabaseAdapterExclusionConstraintError;
class EntityDatabaseAdapterForeignKeyConstraintError extends EntityDatabaseAdapterError {
    state = EntityError_1.EntityErrorState.PERMANENT;
    code = EntityError_1.EntityErrorCode.ERR_ENTITY_DATABASE_ADAPTER_FOREIGN_KEY_CONSTRAINT;
}
exports.EntityDatabaseAdapterForeignKeyConstraintError = EntityDatabaseAdapterForeignKeyConstraintError;
class EntityDatabaseAdapterNotNullConstraintError extends EntityDatabaseAdapterError {
    state = EntityError_1.EntityErrorState.PERMANENT;
    code = EntityError_1.EntityErrorCode.ERR_ENTITY_DATABASE_ADAPTER_NOT_NULL_CONSTRAINT;
}
exports.EntityDatabaseAdapterNotNullConstraintError = EntityDatabaseAdapterNotNullConstraintError;
class EntityDatabaseAdapterUniqueConstraintError extends EntityDatabaseAdapterError {
    state = EntityError_1.EntityErrorState.PERMANENT;
    code = EntityError_1.EntityErrorCode.ERR_ENTITY_DATABASE_ADAPTER_UNIQUE_CONSTRAINT;
}
exports.EntityDatabaseAdapterUniqueConstraintError = EntityDatabaseAdapterUniqueConstraintError;
